routes/api/contacts.js

```diff
@@ -1,27 +1,114 @@
-const express = require('express')
+const express = require("express");
 
-const router = express.Router()
+const router = express.Router();
 
-router.get('/', async (req, res, next) => {
-  res.json({ message: 'Hello World!' })
-})
+const Joi = require("joi");
 
-router.get('/:contactId', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+const responseSchema = Joi.object({
+  name: Joi.string().required(),
+  email: Joi.string().email().required(),
+  phone: Joi.string().required(),
+});
 
-router.post('/', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+const {
+  listContacts,
+  getContactById,
+  addContact,
+  removeContact,
+  updateContact,
+} = require("../../models/contacts");
 
-router.delete('/:contactId', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+router.get("/", async (req, res, next) => {
+  try {
+    const contacts = await listContacts();
+    res.json({
+      status: "success",
+      code: 200,
+      data: {
+        contacts,
+      },
+    });
+  } catch (err) {
+    next(err);
+  }
+});
 
-router.put('/:contactId', async (req, res, next) => {
-  res.json({ message: 'template message' })
-})
+router.get("/:contactId", async (req, res, next) => {
+  const { contactId } = req.params;
+  try {
+    const foundContact = await getContactById(contactId);
+    res.json({
+      status: "success",
+      code: 200,
+      data: { foundContact },
+    });
+  } catch (err) {
+    if (err.message === "Contact not found") {
+      res.status(404).json({ message: "Not found" });
+    } else {
+      next(err);
+    }
+  }
+});
 
-module.exports = router
+router.post("/", async (req, res, next) => {
+  try {
+    const { name, email, phone } = req.body;
+    const newContact = await addContact({ name, email, phone });
+    const validatedBody = responseSchema.validate(req.body);
```

Comment:
Walidację chcemy zrobić jako pierwszą i od razu wyrzucić błąd jeśli jest nie poprawna. Teraz nawet jeśli walidacja wyrzuci błąd, to nadal addContact się wykona ponieważ jest wyżej. dodatkowo przy res trzeba nadać status 201 ponieważ aktualnie zwracasz 200 ( pole w body to nie to samo co status odpowiedzi). Oraz przy 400 możemy wyrzucić błąd przekazywany przez JOI wtedy będzie wiadomo które pole jest niepoprawne.

```
try {
const validatedBody = responseSchema.validate(req.body);
   if (validatedBody.error) {
      return res.status(400).json({ message: validatedBody.error.message });
   }
   const newContact = await addContact(req.body);
   res.status(201).json({
      status: "success",
      code: 201,
      data: { newContact },
    });
} catch (error) {
  console.error(error);
  res.status(500).json({ error: "Unknow error occured." })
}
```

I najlepiej w catch nie rób next(error) tylko tak jak tu pokazałem res.status(500)